### ISSUE TITLE
Unseen Categories Using Categorical Re-coder

### ISSUE THREAD

**Issue by Marcrb2** [dmlc/xgboost#11826]
I’m using the latest version (3.1.1) and testing the new Categorical Re-coder feature in Python. During prediction, I encountered the following error:

Found a category not in the training set.

In previous versions, I handled this by using pd.Categorical and manually storing the categories during training. At inference time, any unseen category was automatically converted to NaN, which allowed the model to process it without failing.

With the new Categorical Re-coder in 3.1, if an unseen category appears during inference, the process fails with the above error.

Is this the intended behavior for the new Categorical Re-coder?
If so, what is the recommended approach for handling unseen categories during inference?

**Comment by trivialfis** [comment]
Hi, thank you for raising the issue

> Is this the intended behavior for the new Categorical Re-coder?

Yes, this is the intended behaviour.

>  If so, what is the recommended approach for handling unseen categories during inference?

Handle it in the pre-processing ETL pipeline as you have been doing.

**Comment by trivialfis** [comment]
The re-coder should handle missing values properly https://github.com/dmlc/xgboost/blob/4bda887974dbde9315e970576197e77a71005cdb/python-package/xgboost/testing/ordinal.py#L237
Please let me know if it doesn't work for you

**Comment by owenprough-sift** [comment]
The documentation on automatic recoding ([link](https://xgboost.readthedocs.io/en/stable/tutorials/categorical.html#auto-recoding-data-consistency)) doesn't mention that category values at inference time must match category values at training. I also assumed that "unknown" values at inference would be ignored or automatically converted to missing/null, therefore allowing me to still predict on records with new/unknown-at-training categorical values. Perhaps the documentation can be updated to clarify this important constraint.

**Comment by trivialfis** [comment]
I will update the document.

In general, XGBoost doesn't silently ignore something unexpected if it can be efficiently checked.

As for unseen categories, it's not a constraint from my perspective as XGBoost doesn't know how to handle it. It's not missing value as defined by XGBoost. Rather it's a completely different type of feature when a new category is added. It's not data drift either since it's a new data type.

There are many ways to handle it. For example, if you know in advance that such thing could happen, you might add an "unknown" category during training and inject some samples with this category as missing value. You can do what you have been doing, simply treat it as missing a inference time. Also, you might consider the category similar to an existing one based on your domain knowledge, and map to that category instead. Alternatively, you can retrain the model, which makes sense as you have a new type of data, consider it as the case where one has a new feature that's not seen during training, you would retrain the model if you want it to properly handle that feature. Silently ignore it would be the last option to consider.

**Comment by seanv507** [comment]
So I come at this from a linear models perspective, where I would say the standard approach would be that unknown categories are mapped to zero (so NOOP), so the idea would be that you instead fall back on other features ( eg if you have age, gender, income and job, as features and  a new, unknown job type in validation/test data), then you only use non job features in the prediction)

So my expectation would be that xgboost would just drop trees that used that categorical split.  This should be an explicit option, I agree that raising an error should be the default.

As far as I am aware, I cannot achieve this desired behaviour from preprocessing the data. Is this something you would consider, or does xgboost tend to mix features too much so that there would be no trees left?

**Comment by trivialfis** [comment]
> So my expectation would be that xgboost would just drop trees that used that categorical split. 

I don't think this is feasible for boosting models. Taking out trees in the middle could be detrimental to accuracy.

**Comment by seanv507** [comment]
My point is that right now xgboost refuses to output any value, which is unusable (ie completely inaccurate).

So taking out trees in the middle *may*  be the the most accurate option.

Another alternative would be to use the mean value for unseen categories, as is done for target encoding (eg scikit-learn, https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.TargetEncoder.html )

>[TargetEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.TargetEncoder.html#sklearn.preprocessing.TargetEncoder) considers missing values, such as np.nan or None, as another category and encodes them like any other category. Categories that are not seen during [fit](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.TargetEncoder.html#sklearn.preprocessing.TargetEncoder.fit) are encoded with the target mean, i.e. target_mean_.

ie during partitioning, the mean value ( ie unpartitioned) value is recorded and used for unseen categories.

**Comment by trivialfis** [comment]
> My point is that right now xgboost refuses to output any value, which is unusable (ie completely inaccurate).

I would like to argue that refusing to output value is the right approach for production data. This requires the data pipeline to handle the data correctly and explicitly. No blind spot.

> ie during partitioning, the mean value ( ie unpartitioned) value is recorded and used for unseen categories.

Seems quite easy to do in the pre-processing stage?

**Comment by seanv507** [comment]
We are agreed:
>> I agree that raising an error should be the default.

I am just asking for an  option.

Afaik, one cannot do this at preprocessing stage, because the optimal partition is selected at each node split as lightgbm.

Therefore  the mean used would be on the residual, not the global mean available at preprocessing time... 


https://lightgbm.readthedocs.io/en/latest/Features.html#optimal-split-for-categorical-features
 
 >>The basic idea is to sort the categories according to the training objective at each split. More specifically, LightGBM sorts the histogram (for a categorical feature) according to its accumulated values (sum_gradient / sum_hessian) and then finds the best split on the sorted histogram.

**Comment by trivialfis** [comment]
Hmm, there's no mean value for categorical features, the mean of `["happy", "sad"]` is `0.5` given its encoding, which doesn't make much sense semantically, nor from a gradient boosting standpoint. Thus far, we have following ideas for ad-hoc handling:
- Using a mean value within split, which might be changed to things like the majority category.
- Drop trees.
- Random choice following a uniform distribution.
- Treat as missing, which is just yet another category.

Honestly, I don't understand why you want XGBoost to handle it because none of the above options is actually "good" in the context of machine learning. We are just patching the model based on no prior information. The majority of the categories has nothing to do with the new unseen category. Actually, choosing the minority category might make sense (which doesn't, but let's just say that it does) since it's likely that value has the least amount of gradient to contribute, close to this unseen new category.

From a (possibly naive) practitioner's perspective, if a category can be ingested into a tabular form of feature, then the pipeline is aware of that category and can make a more informed choice than XGBoost.

Re-opening since this has come up multiple times, I think it's worth investigating and see if we can borrow some ideas from other algorithms like target encoding.

**Comment by seanv507** [comment]
Thank you!

So imo

Optimal partitioning relies on target  encoding during the node split (or
can't be done in preprocessing)

On the other hand, there is also a natural progression between no data
(unseen category) and small amounts of data (rare category), where one
would naturally smooth between the category estimate and the mean (prior to
that node split)

Just as is implemented in sklearns target encoding

https://scikit-learn.org/stable/modules/preprocessing.html#binary-classification-targets


So I would argue it is not an ad-hoc choice to handle unseen categories and
rare categories in a consistent fashion.

Whenever I talk about mean of categorical features I am talking about the
mean of the target value etc.

**Comment by RAMitchell** [comment]
I've done some research on this and I believe there is no satisfying way we can guess unseen categories in XGBoost, at least not without major (breaking) redesign to our tree format and the information we store in it.

Dropping trees just makes boosting not work properly - you might also drop the entire ensemble if the feature is often used.

Using the majority category is biased and we don't store that information.

We could consider it missing but in the likely case there are no missing values in the training data it would always just take the left branch.

You need to handle it yourselves I think.

**Comment by seanv507** [comment]
The catboost authors argue against optimal partitioning at each node, and just using the target statistics (so can be done also in preprocessing).

> In LightGBM [20], categorical features are converted to gradient statistics at each step of gradient
> boosting. Though providing important information for building a tree, this approach can dramatically
> increase (i) computation time, since it calculates statistics for each categorical value at each step, and
> (ii) memory consumption to store which category belongs to which node for each split based on a
> categorical feature. To overcome this issue, LightGBM groups tail categories into one cluster [21] and
> thus looses part of information. Besides, the authors claim that it is still better to convert categorical
> features with high cardinality to numerical features [19]. Note that TS features require calculating
> and storing only one number per one category.
> Thus, using TS as new numerical features seems to be the most efficient method of handling
> categorical features with minimum information loss.
> 

https://arxiv.org/pdf/1706.09516

I would suggest using the target statistics for each new tree (ie looking at residual/gradient) would not be as computationally demanding as per node, but better than using target statistics of the original target variable.


Is this something that has already been considered?

(unknown categories would then be mapped to the mean residual, which is zero?)

**Comment by trivialfis** [comment]
The computation time and memory are not really an issue, trees are usually much smaller than the data (otherwise we would be over fitting), and we are using histogram for evaluation. The argument is mostly about accuracy.

We looked into it before and haven't ruled out an implementation in XGBoost. However, since sklearn has target encoding implementation, one might simply use it instead of building it into the XGBoost library.
